llvm/lib/Target/DirectX/DXILLegalizePass.cpp

```diff
@@ -0,0 +1,209 @@
+//===- DXILLegalizePass.cpp - Legalizes llvm IR for DXIL ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===---------------------------------------------------------------------===//
+//===---------------------------------------------------------------------===//
+///
+/// \file This file contains a pass to remove i8 truncations and i64 extract
+/// and insert elements.
+///
+//===----------------------------------------------------------------------===//
+#include "DXILLegalizePass.h"
+#include "DirectX.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/Pass.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include <functional>
+#include <map>
+#include <stack>
+#include <vector>
+
+#define DEBUG_TYPE "dxil-legalize"
+
+using namespace llvm;
+namespace {
+
+static void fixI8TruncUseChain(Instruction &I,
```

Comment:
I find the nesting in this function hard to follow, especially since the longest case is in the middle. Since all of the instruction types are mutually exclusive, do you think it would read better to structure it like:
```c++
  if (auto *Trunc = dyn_cast<TruncInst>(&I)) {
    // ...
    return;
  }

  if (auto *Cast = dyn_cast<CastInst>(&I)) {
    // ...
    return;
  }

  auto *Cmp = dyn_cast<CmpInst>(&I);
  if (!I.getType()->isIntegerTy(8) &&
      !(Cmp && Cmp->getOperand(0)->getType()->isIntegerTy(8)))
    // Nothing to do...
    return;

  // ...
```

Reply:
I think i have something cleaner let me know in the new pr.